src/diagnostic/emitter.rs

```diff
@@ -0,0 +1,313 @@
+        let mut buffer = StyledBuffer::new();
+
+        // **level:**
+        buffer.puts(diag.level.as_styled_string());
+
+        // level: **message**
+        buffer.puts(StyledString::new(
+            format!(": {}\n", &diag.message),
+            super::styled::Style::MainHeaderMsg,
+        ));
+
+        let max_spaces = self.max_line_num_width(diag);
+        let mut is_first = true;
+
+        if let Some(primary) = diag.primary {
+            self.emit_line(&mut buffer, primary, None, true, max_spaces, diag.level);
+            is_first = false;
+        }
+
+        for (span, label) in diag.spans.iter() {
+            self.emit_line(
+                &mut buffer,
+                *span,
+                Some(label.clone()),
+                is_first,
+                max_spaces,
+                diag.level,
+            );
+            is_first = false;
+        }
+
+        for subd in diag.children.iter() {
+            if let Some(span) = subd.span {
+                buffer.puts(subd.level.as_styled_string());
+
+                buffer.puts(StyledString::new(
+                    format!(": {}\n", subd.message),
+                    super::styled::Style::MainHeaderMsg,
+                ));
+
+                self.emit_line(&mut buffer, span, None, true, max_spaces, subd.level);
+            } else {
+                buffer.puts(StyledString::new(
+                    format!("{:spaces$} = ", "", spaces = max_spaces),
+                    super::styled::Style::LineAndColumn,
+                ));
+
+                buffer.puts(StyledString::new(
+                    format!("{}: ", subd.level.to_str()),
+                    super::styled::Style::MainHeaderMsg,
+                ));
+
+                buffer.puts(StyledString::new(
+                    format!("{}\n", subd.message),
+                    super::styled::Style::NoStyle,
+                ));
+            }
+        }
+
+        buffer.puts(StyledString::new(
+            String::from("\n"),
+            super::styled::Style::NoStyle,
+        ));
+
+        // Render the buffer we have accumulated
+        self.render_buffer(&mut stream, &buffer)?;
+
+        Ok(())
+    }
+
+    fn render_buffer(
+        &self,
+        stream: &mut StandardStream,
+        buffer: &StyledBuffer,
+    ) -> Result<(), Error> {
+        for string in buffer.iter() {
+            let spec = string.style.to_spec();
+
+            if self.colored {
+                stream.set_color(&spec)?;
+            }
+
+            write!(stream, "{}", string.text)?;
+
+            if self.colored {
+                stream.set_color(&ColorSpec::new())?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn max_line_num_width(&self, diag: &Diagnostic) -> usize {
+        let mut spans = Vec::new();
+        let mut max_width = 0;
+
+        if let Some(primary) = diag.primary {
+            spans.push(primary);
+        }
+
+        // That could kind of be a lot of copies, is this a concern?
```

Comment:
I fixed the copy problem, as a reference should be smaller than a &Span. I think overall the switch is a good decision even though theoretically speed isn't as much of a concern when emitting errors because compilation is already usually not going to happen.
Changed in e7b544440bb2fe6b440db04853635364db3274f0